### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by atsheehan**

### CODE REVIEW

File: app/controllers/manufacturers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+class ManufacturersController < ApplicationController
+
+  def index
+    @manufacturers = Manufacturer.all
+  end
+
+  def show
+    @manufacturer = Manufacturer.find(params[:id])
+    @cars = @manufacturer.cars
```
Comment (atsheehan) [inline app/controllers/manufacturers_controller.rb:9]: You don't necessarily need to assign this to a separate instance variable since you can call `@manufacturer.cars` in your view. The less instance variables you require, the easier it is to reuse this view if you ever decide to render it from another controller or action.

File: app/models/car.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+  validates :color, presence: true
+  validates :mileage, presence: true, numericality: {
+            greater_than_or_equal_to: 0
+            }, allow_nil: false
+  validates :year, presence: true, numericality: {
+            greater_than_or_equal_to: 1920,
+            less_than_or_equal_to: (Time.now.year + 1)
+            }, allow_nil: false
+  validates :manufacturer_id, presence: true
```
Comment (atsheehan) [inline app/models/car.rb:12]: You can validate the manufacturer directly which will test that it is an actual `Manufacturer` record rather than just an integer:

``` ruby
validates :manufacturer, presence: true
```

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
   <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track' => true %>
   <%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>
   <%= csrf_meta_tags %>
 </head>
 <body>
 
+<% if flash.any? %>
+  <%= flash[:notice] %>
+  <br>
+<% end %>
```
Comment (atsheehan) [inline app/views/layouts/application.html.erb:14]: Even though you're only using `flash[:notice]` in this case there might be other flash messages you want to display in the future. To make this more generic you can iterate over the flash hash and display each message that exists:

``` HTML+ERB
<% if flash.any? %>
  <% flash.each do |type, message| %>
    <div class="<%= type %>"><%= message %></div>
  <% end %>
<% end %>
```

By wrapping the flash message in a div with a specific class you can style certain flash messages differently (e.g. notice might be blue while alert might be red).

File: app/views/manufacturers/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<%= link_to "Add a New Manufacturer", new_manufacturer_path %>
+
+<br><br>
```
Comment (atsheehan) [inline app/views/manufacturers/index.html.erb:3]: Rather than using break tags everywhere try wrapping different pieces of content with paragraph tags or other divs with classes and IDs. This will give your markup more meaning and allow you to target certain parts for styling later on.

File: spec/controllers/cars_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      car = Car.new
+      get :new, manufacturer_id: @manufacturer
+      expect(assigns(:car).attributes).to eq(car.attributes)
+    end
+    it "renders the :new template" do
+      car = Car.new
+      get :new, manufacturer_id: @manufacturer
+      expect(response).to render_template :new
+    end
+  end
```
Comment (atsheehan) [inline spec/controllers/cars_controller_spec.rb:20]: Although these tests are valid, this behavior is already exercised via the acceptance tests so it is a bit redundant. Controller tests are good when you want to verify a controller returns a certain response code or you want to test against arbitrary (and malicious) user input.

File: spec/controllers/cars_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+          .to_not change(Car, :count)
+      end
+      it "does not save the new car in the database with a negative mileage" do
+        expect{ post :create, manufacturer_id: @manufacturer, car: FactoryGirl.build(:car, mileage: -1).attributes }
+          .to_not change(Car, :count)
+      end
+      it "re-renders the :new template" do
+        post :create, manufacturer_id: @manufacturer, car: FactoryGirl.build(:car, color: nil).attributes
+        expect(response).to render_template :new
+      end
```
Comment (atsheehan) [inline spec/controllers/cars_controller_spec.rb:62]: These tests might be better suited for the model where you can test the validations directly.

File: spec/features/user_adds_new_car_spec.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  scenario "user enters mileage that is negative" do
+    fill_in 'Color', with: @car.color
+    fill_in 'Mileage', with: "-1"
+    fill_in 'Year', with: @car.year
+    fill_in 'Description', with: @car.description
+    click_on 'Create Car'
+
+    expect(page).to have_content("Your Car was not succesfully submitted.")
+  end
+end
```
Comment (atsheehan) [inline spec/features/user_adds_new_car_spec.rb:94]: Acceptance testing can be relatively slow to run so you don't necessarily need to test every edge case. What we're really trying to see is if everything is hooked up correctly so that the app behaves one way when the user successfully creates a car (i.e. they are redirected to a new page) and what happens when there is an error (i.e. they are shown the form).

For more fine-grained control over the validations I would use model-level testing which run much faster.

File: spec/views/cars/new.html.erb_spec.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  it 'displays a field to enter a year' do
+    render
+    expect(rendered).to have_field('Year')
+  end
+
+  it 'displays a field to enter a description' do
+    render
+    expect(rendered).to have_field('Description')
+  end
+end
```
Comment (atsheehan) [inline spec/views/cars/new.html.erb_spec.rb:34]: Although it is good to have thorough test coverage you want to avoid testing redundant code. If this behavior is already tested in the acceptance tests then you're writing duplicate code here. If you ever decide to change how this view behaves you'll have to also change two separate tests.
